packages/core/src/browser/secondary-window-handler.ts

```diff
@@ -0,0 +1,200 @@
+ * Widgets must explicitly implement the `ExtractableWidget` interface to support this.
+ *
+ * This handler manages the opened secondary windows and sets up messaging between them and the Theia main window.
+ * In addition, it provides access to the extracted widgets and provides notifications when widgets are added to or removed from this handler.
+ */
+@injectable()
+export class SecondaryWindowHandler {
+    /** List of currently open secondary windows. Window references should be removed once the window is closed. */
+    protected readonly secondaryWindows: Window[] = [];
+    /** List of widgets in secondary windows. */
+    protected readonly _widgets: Widget[] = [];
+
+    /**
+     * Randomized prefix to be included in opened windows' ids.
+     * This avoids conflicts when creating sub-windows from multiple theia instances (e.g. by opening Theia multiple times in the same browser)
+     */
+    protected readonly prefix = crypto.getRandomValues(new Uint16Array(1))[0];
+
+    protected applicationShell: ApplicationShell;
+
+    protected readonly onDidAddWidgetEmitter = new Emitter<Widget>();
+    /** Subscribe to get notified when a widget is added to this handler, i.e. the widget was moved to an secondary window . */
+    readonly onDidAddWidget = this.onDidAddWidgetEmitter.event;
+
+    protected readonly onDidRemoveWidgetEmitter = new Emitter<Widget>();
+    /** Subscribe to get notified when a widget is removed from this handler, i.e. the widget's window was closed or the widget was disposed. */
+    readonly onDidRemoveWidget = this.onDidRemoveWidgetEmitter.event;
+
+    constructor(@inject(MessageService) protected readonly messageService: MessageService) { }
+
+    /** @returns List of widgets in secondary windows. */
+    get widgets(): ReadonlyArray<Widget> {
+        // Create new array in case the original changes while this is used.
+        return [...this._widgets];
+    }
+
+    /**
+     * Sets up message forwarding from the main window to secondary windows.
+     * Does nothing if this service has already been initialized.
+     *
+     * @param shell The `ApplicationShell` that widgets will be externalized from.
+     */
+    init(shell: ApplicationShell): void {
+        if (this.applicationShell) {
+            // Already initialized
+            return;
+        }
+        this.applicationShell = shell;
+
+        // Set up messaging with secondary windows
+        window.addEventListener('message', (event: MessageEvent) => {
+            console.trace('Message on main window', event);
+            if (event.data.fromExternal) {
+                console.trace('Message comes from secondary window');
+                return;
+            }
+            if (event.data.fromMain) {
+                console.trace('Message has mainWindow marker, therefore ignore it');
+                return;
+            }
+
+            // Filter setImmediate messages. Do not log because these come in with very high frequency.
```

Comment:
```suggestion
            // Filter setImmediate messages. Do not forward because these come in with very high frequency.
```

Can you also explain why these are not needed besides them coming in frequently?